Created Cilium Network Policy and Cilium Cluster Wide Network Policy Test

karina-ranadive opened the pull request:
I created a Cilium Network Policy and Cilium Cluster Wide Network Policy pipeline that will deploy a set number of CCNPS and CNPS that match to pods. It will also scale to a certain number of nodes and scale to a set number of pods. The policies have egress and ingress rules to reflect customer usage and to be able to test control plane resource usage consumption. This test is created based on the existing service churn pipeline framework. 

I created this test to be able to find the cilium cpu, memory, and apiserver usage through testing a 1000 node cilium cluster with endpoints as well as a 3000 node cilium cluster with CES enabled. This is for the upcoming official release of the cnp/ccnp feature. 

Here is the link to my pipeline: https://dev.azure.com/akstelescope/telescope/_build?definitionId=37 and here is a link to a successful run: https://dev.azure.com/akstelescope/telescope/_build/results?buildId=9548&view=results

Since I made changes to files that the other pipelines use, I'm running Service & Cluster Churn:

Regular Pipelines:
- Service Churn: https://dev.azure.com/akstelescope/telescope/_build/results?buildId=9545&view=results
- Cluster Churn: https://dev.azure.com/akstelescope/telescope/_build/results?buildId=9546&view=results

karina-ranadive commented:
> @karina-ranadive please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree company="Microsoft"

anson627 commented on modules/python/clusterloader2/slo/config/cnp_template.yaml:
```diff
@@ -0,0 +1,12 @@
+kind: CiliumNetworkPolicy
+metadata:
+  name: {{.basename}}
+  namespace: slo-1
```
better namespace naming?

karina-ranadive replied:
It's called slo-1 since the pods are deployed in slo-1. I had tried to grab the namespace from load-config directly instead of writing slo-1 but was running into errors that the namespace passed in wasn't matching the object's namespace, so I wrote slo-1 for now since the nginx pods are in the slo-1 namespace and I need the cnp to be in the same namespace.  

<img width="236" alt="image" src="https://github.com/user-attachments/assets/bccca966-7856-489f-883b-03e3793669f9" />

alyssa1303 replied:
maybe we put a comment here to explain why `slo-1` is used and revise later to see if we can pass in namespace

karina-ranadive replied:
sounds good will do that

karina-ranadive replied:
pushed commit that adds comment

anson627 commented on modules/python/clusterloader2/slo/config/load-config.yaml:
```diff
@@ -34,6 +36,12 @@ name: load-config
+# CNP & CCNP Test
+{{$CNPS_PER_NAMESPACE := DefaultParam .CL2_CNPS_PER_NAMESPACE 0}}
+{{$CCNPS := DefaultParam .CL2_CCNPS 0}}
+{{$nameOfNs := "slo-1"}}
```
same here

karina-ranadive replied:
Since I was having issues with the way I tried setting namespace in the picture above, I tried setting it as a string like this and passing it in that way but was also having issues with that. I can remove this since it is not needed.

anson627 commented:
the purpose of test is still exercise k8s control plane, that's why the actual polices are empty, then why do we need nginx  and ping traffic?

jshr-w commented on modules/python/clusterloader2/slo/config/load-config.yaml:
```diff
@@ -108,10 +142,20 @@ steps:
         tuningSet: Sequence
         operationTimeout: {{$operationTimeout}}
+    {{if or $CCNP_TEST $CNP_TEST}}
+        bigDeploymentSize: 0
```
mostly lgtm, thanks Karina! one non-blocking comment, think it may be better to make bigDeploymentSize and bigDeploymentsPerNamespace overwritable CL2_ parameters and overwrite them in slo.py in the long run

karina-ranadive replied:
Thanks! This makes sense, will do this in long run

karina-ranadive commented:
Updated Changes:
- Removed nginx pods and now using the container image that was already present
- Added rules to CNP and CCNP template to reflect customer usage to be able to test control plane resource usage consumption

alyssa1303 commented on steps/engine/clusterloader2/slo/execute.yml:
```diff
@@ -13,9 +13,9 @@ steps:
 
       PYTHONPATH=$PYTHONPATH:$(pwd) python3 $PYTHON_SCRIPT_FILE configure \
-        $CPU_PER_NODE $NODE_COUNT $NODE_PER_STEP $MAX_PODS \
+        $CPU_PER_NODE $NODE_COUNT $NODE_PER_STEP $MAX_PODS ${PODS_IN_NODE:-0} \
```
Instead of this new pods_in_node, I'd suggest we use max_pods to define how many pods to schedule since the variable already exists

alyssa1303 commented on modules/python/clusterloader2/slo/slo.py:
```diff
@@ -24,14 +24,16 @@
     throughput = 100
     nodes_per_namespace = min(node_count, DEFAULT_NODES_PER_NAMESPACE)
 
     pods_per_node = DEFAULT_PODS_PER_NODE
     if service_test:
         pods_per_node = LOAD_PODS_PER_NODE
 
+    if cnp_test or ccnp_test:
+        pods_per_node = pods_in_node
```
Looks like we want to customize the number of pod per nodes. Let rewrite this part to use max_pods variables instead adding new variables for different tests. This is very confusing.

karina-ranadive replied:
fixed with commit "removed pods_in_node and utilizing max_pods", thanks!

alyssa1303 commented on steps/engine/clusterloader2/slo/collect.yml:
```diff
@@ -16,8 +16,8 @@ steps:
     PYTHONPATH=$PYTHONPATH:$(pwd) python3 $PYTHON_SCRIPT_FILE collect \
-      $CPU_PER_NODE $NODE_COUNT $MAX_PODS $REPEATS \
-      $CL2_REPORT_DIR "$CLOUD_INFO" $RUN_ID $RUN_URL $SERVICE_TEST $TEST_RESULTS_FILE \
+      $CPU_PER_NODE $NODE_COUNT ${MAX_PODS:-0} \
```
I think just `$MAX_PODS` is good enough as I believe other pipelines already use the value

karina-ranadive replied:
Oh so I kept it as that in case in the cnp/ccnp pipeline the max pods value is left blank in the variables in the pipeline. In cnp-ccnp-feature.yaml, I take in the variable from the pipeline into max_pods and I had done testing if any of the values are blank, I found that it causes an error when hitting the execute or collect file so I just did it for preventative measure for precaution. Would it be okay to keep it?

alyssa1303 replied:
Yeah I guess it's fine to keep it now. But in the future, we should enforce this value to used by all tests so we don't have different hard-coded value anymore, and failure to set it should result in pipeline failure

karina-ranadive commented:
> the purpose of test is still trying to exercise control plane with crds reconcile, that's why the actual polices are empty, then why do we need nginx and ping traffic?

I responded to the question in chat and made changes to remove nginx pods and removed ping traffic and I am utilizing container image that was already present. I also added rules to the cnp and ccnp template to reflect customer usage to be able to test control plane resource usage consumption.

jshr-w reviewed (approved):
thanks Karina!
